src/create_venv.sh

```diff
@@ -1,136 +1,258 @@
-  awk '{print $1}'<<<"$(sha256sum <<<"${tmp}")"
-}
+[[ "${_YB_CREATE_VENV_INCLUDED:=""}" == "yes" ]] && return 0
+_YB_CREATE_VENV_INCLUDED=yes
+YB_VERBOSE=${YB_VERBOSE:-false}
 
 set -e -u -o pipefail
 
-YB_PYTHON_VERSION=${YB_PYTHON_VERSION:-3.7}
-py_major_version=$(cut -f1 -d. <<<"${YB_PYTHON_VERSION}")
+_src_dir="${BASH_SOURCE%/*}"
+if [[ ! -d "${_src_dir}" ]]; then _src_dir="$PWD"; fi
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/logger.sh
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/os.sh
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/detect_python.sh
 
+# -------------------------------------------------------------------------------------------------
+# Global variables used in this module
+# -------------------------------------------------------------------------------------------------
+# This preserves the current existing behavior of assuming requirements_frozen.txt is always
+# up to date.  TODO: default this to true to ensure frozen file is 'correct'
+YB_BUILD_STRICT=${YB_BUILD_STRICT:-false}
+# Set this to true to force recreation of requirements_frozen.txt
+YB_RECREATE_VIRTUALENV=${YB_RECREATE_VIRTUALENV:-false}
+# New-style VENV base dir, only used when YB_PUT_VENV_IN_PROJECT_DIR is false
 YB_VENV_BASE_DIR=${YB_VENV_BASE_DIR:-~/.venv/yb}
-YB_REFRESH_VENV=${YB_REFRESH_VENV:-false}
+# This preserves the current existing behavior of putting the VENV in the same directory
+# as requirements.txt
+YB_PUT_VENV_IN_PROJECT_DIR=${YB_PUT_VENV_IN_PROJECT_DIR:-true}
 
-echo "Using YB_PYTHON_VERSION=${YB_PYTHON_VERSION}"
+verbose "Using YB_PYTHON_VERSION=${YB_PYTHON_VERSION}"
+# shellcheck disable=SC2154
+verbose "Using ${yb_python_interpreter} (${yb_python_version_actual})"
 
-set +u
-if [[ -n "${VIRTUAL_ENV}" ]] && [[ -d "${VIRTUAL_ENV}" ]]; then
-  echo "Please exit your current venv first."
-  echo "type 'deactivate' in your shell"
-  exit 1
-fi
-set -u
-
-py_cmd=''
-for cmd in "python${YB_PYTHON_VERSION}" "python${py_major_version}" "python"; do
-  if cmd="$(command -v "${cmd}")"; then
-    if "${cmd}" --version 2>&1 | grep "${YB_PYTHON_VERSION}"; then
-      py_cmd="${cmd}"
-      break
+# -------------------------------------------------------------------------------------------------
+# Internal functions used in this module.  These shouldn't be called directly outside this module.
+# -------------------------------------------------------------------------------------------------
+
+function ybcv::text_file_sha_ignore_comments() {
+  local file=$1
+  local tmp
+  tmp="$(sort -u <<<"$(grep -Ev '^[[:space:]]*#' "${file}")")"
+  # shellcheck disable=SC2154
+  awk '{print $1}'<<<"$(${yb_sha256sum} <<<"${tmp}")"
+}
+
+function ybcv::needs_refreeze() {
+  local reqs_sha=$1
+  local frozen_file=$2
+  if [[ -f "${frozen_file}" ]]; then
+    if ! grep "# YB_SHA: ${reqs_sha}" "${frozen_file}" >/dev/null 2>&1; then
+      return 0
     fi
+  else
+    return 0
   fi
-done
+  # shellcheck disable=SC2086
+  false
```

Comment:
Let's not call an external command here just to return a non-zero exit code.
```suggestion
  return 1
```
Or maybe we could create constants in some global library:
```
declare -i -r YB_TRUE_RET_VAL=0
declare -i -r YB_FALSE_RET_VAL=1
```